tools/src/test/java/org/apache/kafka/tools/GetOffsetShellTest.java

```diff
@@ -391,4 +420,15 @@ private String[] addBootstrapServer(String... args) {
 
         return newArgs.toArray(new String[0]);
     }
+
+    private void retryUntilEqual(List<Row> expected, Supplier<List<Row>> outputSupplier) {
+        try {
+            TestUtils.waitForCondition(
+                    () -> expected.equals(outputSupplier.get()),
+                    "TopicOffsets did not match. Expected: " + expectedTestTopicOffsets() + ", but was: " + outputSupplier.get()
```

Comment:
Could you use `admin` to list latest offsets after it fails? for example:

```java
    private Set<Row> parse() throws ExecutionException, InterruptedException {
        try (Admin admin = cluster.createAdminClient()) {
            Set<String> topics = admin.listTopics(new ListTopicsOptions().listInternal(true)).listings().get()
                    .stream().map(TopicListing::name).collect(Collectors.toSet());
            Map<TopicPartition, OffsetSpec> offsetRequest = admin.describeTopics(topics)
                    .allTopicNames().get().entrySet().stream().flatMap(entry -> entry.getValue().partitions()
                            .stream().map(p -> new AbstractMap.SimpleImmutableEntry<>(new TopicPartition(entry.getKey(), p.partition()), OffsetSpec.latest())))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
            return admin.listOffsets(offsetRequest).all().get().entrySet().stream()
                    .map(entry -> new Row(entry.getKey().topic(), entry.getKey().partition(), entry.getValue().offset()))
                    .collect(Collectors.toSet());
        }
    }
```

Reply:
So sth like below ? Print out the final offset as error msg for debugging ?
```
TestUtils.waitForCondition(
                    () -> expected.equals(outputSupplier.get()),
                    "TopicOffsets did not match. Expected: " + expectedTestTopicOffsets() + ", but was: "
                            + outputSupplier.get() + ". Final offsets: " + getFinalOffsets()
            );
```

Reply:
yep, I loop the test with this PR 1000 times, all pass :(

It seems we need to reproduce the failure on our CI

Reply:
Ok done. Let's see what we get from there =)

Reply:
I have the same error in another PR (https://github.com/apache/kafka/pull/15621) :(